merkle/hstar2.go

```diff
@@ -0,0 +1,124 @@
+package merkle
+
+import (
+	"fmt"
+	"math/big"
+	"sort"
+
+	"github.com/google/trillian"
+)
+
+// HStar2Value represents a leaf for the HStar2 sparse merkle tree
+// implementation.
+type HStar2Value struct {
+	// TODO(al): remove big.Int
+	Index *big.Int
+	Value trillian.Hash
+}
+
+// HStar2 is a recursive implementation for calulating the root hash of a sparse
+// merkle tree.
+type HStar2 struct {
+	hasher          TreeHasher
+	hStarEmptyCache []trillian.Hash
+}
+
+// NewHStar2 creates a new HStar2 tree calculator based on the passed in
+// TreeHasher.
+func NewHStar2(treeHasher TreeHasher) HStar2 {
+	return HStar2{
+		hasher:          treeHasher,
+		hStarEmptyCache: []trillian.Hash{treeHasher.HashLeaf([]byte(""))},
+	}
+}
+
+// HStar2 calculates the root of a sparse merkle tree of depth n which contains
+// the given set of non-null leaves.
+func (s *HStar2) HStar2(n int, values []HStar2Value) (trillian.Hash, error) {
+	by(indexLess).Sort(values)
+	offset := big.NewInt(0)
+	return s.hStar2b(n, values, offset)
+}
+
+// hStarEmpty calculates (and caches) the "null-hash" for the requested tree
+// level.
+func (s *HStar2) hStarEmpty(n int) (trillian.Hash, error) {
+	if len(s.hStarEmptyCache) <= n {
+		emptyRoot, err := s.hStarEmpty(n - 1)
+		if err != nil {
+			return nil, err
+		}
+		h := s.hasher.HashChildren(emptyRoot, emptyRoot)
+		if len(s.hStarEmptyCache) != n {
+			return nil, fmt.Errorf("cache wrong size - expected %d, but cache contains %d entries", n, len(s.hStarEmptyCache))
+		}
+		s.hStarEmptyCache = append(s.hStarEmptyCache, h)
+	}
+	if n >= len(s.hStarEmptyCache) {
+		return nil, fmt.Errorf("cache wrong size - expected %d or more, but cache contains %d entries", n, len(s.hStarEmptyCache))
```

Comment:
I might be misreading it but the error message and if condition don't seem to match up.